Indexer locking rework

ifedorenko opened the pull request.

cstamas committed:
NEXUS-5250 Updated to use maven indexer 5.0.0

ifedorenko committed:
NEXUS-5250 Reworked lucene index locking implementation

See DefaultIndexerManager javadoc for more details.

Known problems and outstanding issues.

Possible reordering of add/remove item requests because all index
update requests are processed indipentently by background threads.

Reindex local storage and download full remote index loose
concurrent index changes.

Unnecessary exclusive lock around incremental local storage scan.
Required because maven indexer always scans to a temporary context
first, them performs target.replace(temporary). Should be trivial
to fix in maven indexer but need to understand why temporary
context was originally implemented.

Inconsistent proxy repository index. When items are removed from
local storage they are removed from the index. Redownloading full
index will make the items reappear.

Signed-off-by: Igor Fedorenko <igor@ifedorenko.com>

ifedorenko committed:
NEXUS-5250 Removed now unnecessary search-retry logic

... also made search iterator close() invocation more apparent.

Signed-off-by: Igor Fedorenko <igor@ifedorenko.com>

jdillon commented on nexus/nexus-core-plugins/nexus-indexer-lucene-plugin-parent/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/DefaultIndexerManager.java:
```diff
@@ -503,19 +558,21 @@ public void setRepositoryIndexContextSearchable( String repositoryId, boolean se
+                {
+                    getLogger().debug( "Searching on repository ID='" + repositoryId + "' is set to: "
+                                           + String.valueOf( searchable ) );
+                }
```
This can/should use {} syntax an avoid the guard.  String.valueOf( searchable ) is superfluous.

jdillon commented on nexus/nexus-core-plugins/nexus-indexer-lucene-plugin-parent/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/DefaultIndexerManager.java:
```diff
@@ -593,82 +650,92 @@ public void addItemToIndex( Repository repository, StorageItem item )
-                            ac = artifactContextProducer.getArtifactContext( context, file );
-                        }
-                        catch ( IllegalArgumentException e )
+                        if ( getLogger().isDebugEnabled() )
```
pointless guard

jdillon commented on nexus/nexus-core-plugins/nexus-indexer-lucene-plugin-parent/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/DefaultIndexerManager.java:
```diff
@@ -698,82 +765,91 @@ public void removeItemFromIndex( Repository repository, StorageItem item )
+            getLogger().debug(
+                "Deleting artifact " + ai.groupId + ":" + ai.artifactId + ":" + ai.version
+                    + " from index (DELETE)." );
+        }
```
more overly complex logging, don't guard, use {}

jdillon commented on nexus/nexus-core-plugins/nexus-indexer-lucene-plugin-parent/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/DefaultIndexerManager.java:
```diff
@@ -1193,11 +1258,65 @@ public InputStream retrieve( String name )
+            {
+                getLogger().info(
+                    RepositoryStringUtils.getFormattedMessage(
+                        "Remote indexes updated successfully for repository %s", repository ) );
```
this is pretty crazy IMO, this ends up doing 2 String.format operations.  1 of these can be avoided by using {} properly.

jdillon commented on nexus/nexus-core-plugins/nexus-indexer-lucene-plugin-parent/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/index/DefaultIndexerManager.java:
```diff
@@ -2270,9 +2210,17 @@ public IteratorSearchResponse searchArtifactSha1ChecksumIterator( String sha1Che
-            getLogger().error( "Got I/O exception while searching for query \"" + bq.toString() + "\"", e );
+            lockedContexts.lock.unlock();
+
+            getLogger().error( "Got I/O exception while searching for query \"" + req.getQuery().toString() + "\"", e );
```
uuuuuh... a) should use {}, b) you don't need to toString() when string concatenating... its redundant.

jdillon commented on nexus/nexus-core-plugins/nexus-indexer-lucene-plugin-parent/nexus-indexer-lucene-plugin/src/main/java/org/sonatype/nexus/rest/indexng/SearchNGIndexPlexusResource.java:
```diff
@@ -509,288 +461,281 @@ protected void repackIteratorSearchResponse( Request request, Map<String, String
+                    + " unique GA records. Lucene scored documents first=" + firstDocumentScore + ", last="
+                    + lastDocumentScore + ". Main processing loop took "
+                    + ( System.currentTimeMillis() - startedAtMillis ) + " ms." );
+
```
of all the things to guard, this would be one valid use of guarding...

ifedorenko commented:
I agree logging needs cleanup, and we need get rid of evil Slf4jPlexusLogger too, but I am sure if this should be part of locking rework pull request or done separately. I all see how it goes.

jdillon commented:
I've nuked the Slf4jPlexusLogger already, so if these bits were in here because of that, rebase/merge master and use the proper slf4j api.

ifedorenko committed:
NEXUS-5250 Reindex local storage should not block index queries

Proper fix requires new/updated maven indexer API as per
http://jira.codehaus.org/browse/MINDEXER-65. As a workaround,
copied DefaultScannerListener to make it direclty usable in
nexus.

Signed-off-by: Igor Fedorenko <igor@ifedorenko.com>

cstamas commented:
+1

ifedorenko commented:
I started feature build https://builds.sonatype.org/view/nexus-features/job/nexus-oss-its-feature/497/, will merge this pull request if successful.
